File: frontend/src/rql/lexer.ts
import { lexError, LexError, Result } from "./common";

export enum TokenKind {
  Ident = "Ident",
  StringLit = "StringLit",
  NumberLit = "NumberLit",
  DateLit = "DateLit",
  NullLit = "Null",
  TrueLit = "True",
  FalseLit = "False",
  // special
  Bar = "|",
  LParen = "(",
  RParen = ")",
  Comma = ",",
  // operators
  OpNot = "!",
  OpPlus = "+",
  OpMinus = "-",
  OpTimes = "*",
  OpDivide = "/",
  OpAssign = "=",
  OpEqual = "==",
  OpNotEqual = "!=",
  OpLess = "<",
  OpGreater = ">",
  OpLessEq = "<=",
  OpGreaterEq = ">=",
  OpContains = "contains",
  OpNotContains = "!contains",
  OpAnd = "and",
  OpOr = "or",
}

export interface Token {
  kind: TokenKind;
  value: string;
  pos: number;
}

export interface Lexer {
  input: string;
  pos: number;

  tokens: Token[];
  error: LexError | null;
};

export type LexResult = Result<
  { tokens: Token[] },
  { error: LexError }
>;


function makeToken(kind: TokenKind, value: string, pos?: number) {
  return { kind, value, pos: pos ? pos : -1 };
}

function isWhitespace(c: string): boolean { return (c === ' ') || (c === '\t') || (c === '\n') || (c === '\r'); }
function isAlphabetical(c: string): boolean { return (('A' <= c) && (c <= 'Z')) || (('a' <= c) && (c <= 'z')); }
function isDigit(c: string): boolean { return ('0' <= c) && (c <= '9'); }
function isAlphaNumeric(c: string): boolean { return isAlphabetical(c) || isDigit(c); }
function isIdent(c: string): boolean { return isAlphaNumeric(c) || (c === '_'); }

function inputLeft(lexer: Lexer): boolean { return (lexer.pos < lexer.input.length); }
function hasError(lexer: Lexer): boolean { return (lexer.error !== null); }
function currentChar(lexer: Lexer): string { return lexer.input[lexer.pos]; }
function peekChar(lexer: Lexer): string | undefined { return (lexer.input.length > lexer.pos + 1) ? lexer.input[lexer.pos + 1] : undefined; }

function lexicalError(lexer: Lexer, error: string) {
  lexer.error = lexError(error, lexer.input, lexer.pos);
}

function skipMultilineComment(lexer: Lexer) {
  const startPos = lexer.pos;
  advance(lexer); // skip /
  advance(lexer); // skip *
  while (inputLeft(lexer)) {
    if (currentChar(lexer) === '*' && peekChar(lexer) === '/') {
      break;
    }
    advance(lexer);
  }
  if (!inputLeft(lexer)) {
    lexer.error = lexError("Un treminated multiline comment", lexer.input, startPos);
  } else {
    advance(lexer); // skip *
    advance(lexer); // skip /
  }
}

function skipSingleLineComment(lexer: Lexer) {
  advance(lexer); // skip /
  advance(lexer); // skip /
  while (inputLeft(lexer) && currentChar(lexer) !== '\n') {
    advance(lexer);
  }
  if (inputLeft(lexer)) advance(lexer); // skip newline
}

function skipWhitespace(lexer: Lexer): boolean {
  let skipped = false;
  while (inputLeft(lexer)) {
    if (isWhitespace(currentChar(lexer))) {
      skipped = true;
      advance(lexer);
      continue;
    }
    break;
  }
  return skipped;
}

function addToken(lexer: Lexer, token: Token) {
  lexer.tokens.push({ ...token, pos: token.pos === -1 ? lexer.pos : token.pos } );
}

function addTokenSubstring(lexer: Lexer, startPos: number, kind: TokenKind) {
  const token = makeToken(kind, lexer.input.substring(startPos, lexer.pos), startPos);
  addToken(lexer, token);
}

function advance(lexer: Lexer) { lexer.pos++; }
function advanceWithToken(lexer: Lexer, token: Token) {
  advance(lexer);
  addToken(lexer, token);
}
function accept(lexer: Lexer, c: string, token?: Token): boolean {
  if (inputLeft(lexer) && currentChar(lexer) === c) {
    if (token !== undefined) {
      advanceWithToken(lexer, token);
    } else {
      advance(lexer);
    }
    return true;
  }
  return false;
}

function acceptString(lexer: Lexer, s: string, token?: Token): boolean {
  const start = lexer.pos;
  let i = 0;
  while (inputLeft(lexer) && currentChar(lexer) === s.charAt(i)) {
    advance(lexer);
    i++;
  }
  if (i === s.length && (!inputLeft(lexer) || !isIdent(currentChar(lexer)))) {
    if (token) addToken(lexer, token);
    return true;
  } else {
    lexer.pos = start;
    return false;
  }
}

function lex(lexer: Lexer) {
  switch (currentChar(lexer)) {
    case ' ': case '\t': case '\n': case '\r':
      skipWhitespace(lexer);
      break;
    case '|':
      advanceWithToken(lexer, makeToken(TokenKind.Bar, "|"));
      break;
    case '(':
      advanceWithToken(lexer, makeToken(TokenKind.LParen, "("));
      break;
    case ')':
      advanceWithToken(lexer, makeToken(TokenKind.RParen, ")"));
      break;
    case ',':
      advanceWithToken(lexer, makeToken(TokenKind.Comma, ","));
      break;

    case '!':
      advance(lexer);
      if (inputLeft(lexer) && currentChar(lexer) === '=') {
        advanceWithToken(lexer, makeToken(TokenKind.OpNotEqual, "!="));
      } else if (acceptString(lexer, "contains", makeToken(TokenKind.OpNotContains, "!contains"))) {
        break;
      } else {
        addToken(lexer, makeToken(TokenKind.OpNot, "!"));
      }
      break;
    case '+':
      advanceWithToken(lexer, makeToken(TokenKind.OpPlus, "+"));
      break;
    case '-':
      advanceWithToken(lexer, makeToken(TokenKind.OpPlus, "-"));
      break;
    case '*':
      advanceWithToken(lexer, makeToken(TokenKind.OpTimes, "*"));
      break;
    case '/':
      if (peekChar(lexer) === '/') {
        skipSingleLineComment(lexer);
        break;
      } else if (peekChar(lexer) === '*') {
        skipMultilineComment(lexer);
        break;
      }
      advanceWithToken(lexer, makeToken(TokenKind.OpDivide, "/"));
      break;
    case '=': {
      advance(lexer);
      if (inputLeft(lexer) && currentChar(lexer) === '=') {
        advanceWithToken(lexer, makeToken(TokenKind.OpEqual, "=="));
      } else {
        addToken(lexer, makeToken(TokenKind.OpAssign, "="));
      }
      break;
    }
    case '<': {
      advance(lexer);
      if (inputLeft(lexer) && currentChar(lexer) === '=') {
        advanceWithToken(lexer, makeToken(TokenKind.OpLessEq, "<="))
      } else {
        addToken(lexer, makeToken(TokenKind.OpLess, "<"));
      }
      break;
    }
    case '>': {
      advance(lexer);
      if (inputLeft(lexer) && currentChar(lexer) === '=') {
        advanceWithToken(lexer, makeToken(TokenKind.OpGreaterEq, ">="))
      } else {
        addToken(lexer, makeToken(TokenKind.OpGreater, ">"));
      }
      break;
    }

  //StringLit = "StringLit",
    case '"': {
      advance(lexer);
      const start = lexer.pos;
      while (inputLeft(lexer)) {
        if (currentChar(lexer) === '\\') {
          advance(lexer);
          if (inputLeft(lexer)) {
            advance(lexer);
          }
          continue;
        } else if (currentChar(lexer) === '"') {
          break;
        }
        advance(lexer);
      }

      addTokenSubstring(lexer, start, TokenKind.StringLit);
      advance(lexer); // skip the ending "
      break;
    }
    case "'":
      advance(lexer);
      const start = lexer.pos;
      while (inputLeft(lexer)) {
        if (currentChar(lexer) === '\\') {
          advance(lexer);
          if (inputLeft(lexer)) {
            advance(lexer);
          }
          continue;
        } else if (currentChar(lexer) === "'") {
          break;
        }
        advance(lexer);
      }

      addTokenSubstring(lexer, start, TokenKind.StringLit);
      advance(lexer); // skip the ending '
      break;

    case '0': case '1': case '2': case '3': case '4': case '5': case '6': case '7': case '8': case '9': {
      const startPos = lexer.pos;
      let digits = 1;
      advance(lexer);
      while (inputLeft(lexer) && isDigit(currentChar(lexer))) {
        digits++;
        advance(lexer);
      }
      if (digits === 4 && inputLeft(lexer) && accept(lexer, '-')) {
        // Date literal
        const zero = accept(lexer, '0');
        const one = !zero && accept(lexer, '1');
        if (!(zero || one)) {
          lexicalError(lexer, "Invalid date literal: '0' or '1' expected");
          return;
        }
        if (!inputLeft(lexer) || !isDigit(currentChar(lexer))) {
          lexicalError(lexer, "Invalid date literal: digit expected");
          return;
        } else if (zero && currentChar(lexer) === '0') {
          lexicalError(lexer, "Invalid date literal: '00' is not a valid month");
          return;
        } else if (one && !['0', '1', '2'].includes(currentChar(lexer))) {
          lexicalError(lexer, "Invalid date literal: '1" + currentChar(lexer) + "' is not a valid month");
          return;
        }
        advance(lexer);
        if (!inputLeft(lexer) || !accept(lexer, '-')) {
          lexicalError(lexer, "Invalid date literal: '-' expected");
          return;
        }
        const d0 = accept(lexer, '0');
        const d1 = !d0 && accept(lexer, '1');
        const d2 = !d0 && !d1 && accept(lexer, '2');
        const d3 = !d0 && !d1 && !d2 && accept(lexer, '3');
        if (!inputLeft(lexer)) {
          lexicalError(lexer, "Invalid date literal: digit expected");
          return;
        } else if (d3 && !['0', '1'].includes(currentChar(lexer))) {
          lexicalError(lexer, "Invalid date literal: '3" + currentChar(lexer) + "' is not a valid day of month");
          return;
        } else if (!isDigit(currentChar(lexer))) {
          lexicalError(lexer, "Invalid date literal: digit expected");
          return;
        }
        advance(lexer);

        if (accept(lexer, 'T')) {
          // TODO: implement
        }

        addTokenSubstring(lexer, startPos, TokenKind.DateLit);
      } else if (inputLeft(lexer) && currentChar(lexer) === '.') {
        // Floating point literal
        advance(lexer);
        if (!inputLeft(lexer) || !isDigit(currentChar(lexer))) {
          lexicalError(lexer, "Invalid floating point literal");
          return;
        }

        while (inputLeft(lexer) && isDigit(currentChar(lexer))) {
          advance(lexer);
        }

        if (accept(lexer, 'e') || accept(lexer, 'E')) {
          accept(lexer, '+') || accept(lexer, '-');
          if (inputLeft(lexer) && isDigit(currentChar(lexer))) {
            while (inputLeft(lexer) && isDigit(currentChar(lexer))) {
              advance(lexer);
            }
            addTokenSubstring(lexer, startPos, TokenKind.NumberLit);
          } else {
            lexicalError(lexer, "Invalid floating point literal, expecting exponent");
            return;
          }
        } else {
          addTokenSubstring(lexer, startPos, TokenKind.NumberLit);
        }
      } else {
        // Integer literal
        addTokenSubstring(lexer, startPos, TokenKind.NumberLit);
      }
      break;
    }

    default: {
      if (acceptString(lexer, "null", makeToken(TokenKind.NullLit, "null"))) {
        break;
      } else if (acceptString(lexer, "true", makeToken(TokenKind.TrueLit, "true"))) {
        break;
      } else if (acceptString(lexer, "false", makeToken(TokenKind.FalseLit, "false"))) {
        break;
      } else if (acceptString(lexer, "and", makeToken(TokenKind.OpAnd, "and"))) {
        break;
      } else if (acceptString(lexer, "or", makeToken(TokenKind.OpOr, "or"))) {
        break;
      } else if (acceptString(lexer, "contains", makeToken(TokenKind.OpContains, "contains"))) {
        break;
      } else if (!isAlphabetical(currentChar(lexer))) {
        lexicalError(lexer, "Invalid character '" + currentChar(lexer) + "'");
        advance(lexer);
        break;
      }
      const start = lexer.pos;
      while (inputLeft(lexer) && isIdent(currentChar(lexer))) {
        advance(lexer);
      }
      addTokenSubstring(lexer, start, TokenKind.Ident);
    }
  }
}

export function tokenize(input: string): LexResult {
  const lexer: Lexer = { input, pos: 0, tokens: [], error: null };
  while (inputLeft(lexer) && !hasError(lexer)) {
    lex(lexer);
  }

  if (lexer.error !== null) {
    return { _type: "failed", error: lexer.error };
  } else {
    return { _type: "success", tokens: lexer.tokens };
  }
}

File: backend/src/main/scala/castool/rql/Lexer.scala
package castool.rql

final case class InputPos(input: String, line: Int, column: Int)

object Lexer {

  case class Input(input: String, pos: Int) {
    def hasInput(): Boolean = pos < input.length()
    def current(): Char = input.charAt(pos)
    def accept(x: Char): Boolean = hasInput() && current() == x
    def advance(): Input = {
      assert(hasInput())
      copy(pos = pos + 1)
    }
    def substringFrom(startPos: Int): String = input.substring(startPos, pos)
  }

  sealed trait State {
    def isValid: Boolean
    def hasInput: Boolean
  }

  final case class Valid(input: Input, tokens: Vector[Token]) extends State {
    override def isValid: Boolean = true
    override def hasInput: Boolean = input.hasInput()
    def advance(): Valid = copy(input = input.advance())
    def emit(token: Token): Valid = copy(tokens = tokens :+ token)
    def reject(message: String): Reject = Reject(this, message)
  }
  final case class Reject(prev: Valid, message: String) extends State {
    override def isValid: Boolean = false
    override def hasInput: Boolean = false
  }

  trait Consumer {
    def accept(state: Valid): State
    def :>(other: Consumer): And = And(this, other)
    def :>(c: Char): And = And(this, AcceptChar(c))
    def :>(s: String): And = And(this, AcceptString(s))
    def :|(other: Consumer): Or = Or(this, other)
    def :|(c: Char): Or = Or(this, AcceptChar(c))
    def :|(s: String): Or = Or(this, AcceptString(s))
    def *(): Consumer = RepeatWhile(this)
    def apply(n: Int): Consumer = Times(this, n)
    def ![T <: Token](fn: (String, Int) => T): Consumer = Emit(this, fn)
    def ![T <: Token](fn: Int => T): Consumer = Emit(this, (_, pos) => fn(pos))
    def !![T <: Token](map: (String, Int) => scala.util.Try[T]) = EmitMap(this, map)
  }

  case class And(a: Consumer, b: Consumer) extends Consumer {
    def accept(state: Valid): State = a.accept(state) match {
      case next: Valid => b.accept(next)
      case r: Reject => r
    }
  }

  case class Or(a: Consumer, b: Consumer) extends Consumer {
    def accept(state: Valid): State = a.accept(state) match {
      case next: Valid => next
      case r: Reject => b.accept(state)
    }
  }

  case class AcceptChar(c: Char) extends Consumer {
    def accept(state: Valid): State = if (state.input.accept(c)) {
      state.advance()
    } else {
      state.reject(s"Expected '$c'")
    }
  }

  case class AcceptAny(c: Iterable[Char]) extends Consumer {
    def accept(state: Valid): State = {
      //println("Accept any " + c)
      var it = c.iterator
      while (it.hasNext) {
        if (state.input.accept(it.next())) {
          return state.advance()
        }
      }
      state.reject(s"Expected any of ${c.map(x => s"'$x'").mkString(", ")}")
    }
  }

  object AcceptAny {
    def apply(c: Char*): AcceptAny = AcceptAny(c)
  }

  case class AcceptAnyBut(c: Iterable[Char]) extends Consumer {
    def accept(state: Valid): State = {
      var it = c.iterator
      while (it.hasNext) {
        if (state.input.accept(it.next())) {
          return state.reject(s"Expected other than ${c.map(x => s"'$x'").mkString(", ")}")
        }
      }
      state.advance()
    }
  }

  object AcceptAnyBut {
    def apply(c: Char*): AcceptAnyBut = AcceptAnyBut(c)
  }

  case class AcceptString(str: String) extends Consumer {
    def accept(state: Valid): State = {
      var index = 0
      var s = state
      while (index < str.length && s.input.accept(str.charAt(index))) {
        s = s.advance()
        index += 1
      }
      if (index == str.length()) {
        s
      } else {
        s.reject(s"Expected '$str'")
      }
    }
  }

  case class Not(c: Consumer) extends Consumer {
    def accept(state: Valid): State = c.accept(state) match {
      case r: Reject => state
      case s: Valid => s.reject(s"Expected other than '${s.tokens.last}'") // TODO: expects that there is at least one token
    }
  }

  object Nop extends Consumer {
    def accept(state: Valid): State = state
  }

  case class Emit[T <: Token](c: Consumer, tokenFn: (String, Int) => T) extends Consumer {
    def accept(state: Valid): State = {
      c.accept(state) match {
        case v: Valid =>
          val startPos = state.input.pos
          val value = v.input.substringFrom(startPos)
          v.emit(tokenFn(value, startPos))
        case _ =>
          state.reject("Unexpected token")
      }
    }
  }

  case class EmitMap[T <: Token](c: Consumer, tokenMap: (String, Int) => scala.util.Try[T]) extends Consumer {
    def accept(state: Valid): State = {
      val next = c.accept(state)
      next match {
        case v: Valid =>
          val startPos = state.input.pos
          val value = v.input.substringFrom(startPos)
          tokenMap(value, startPos) match {
            case scala.util.Success(token) => v.emit(token)
            case scala.util.Failure(ex) => state.reject(s"${ex.getMessage()}")
          }
        case _ =>
          state.reject("Unexpected token")
      }
    }
  }

  case class RepeatWhile(c: Consumer) extends Consumer {
    def accept(state: Valid): State = {
      var s = c.accept(state)
      if (s.isValid) {
        var lastValid = s
        while (s.isValid && s.hasInput) {
          s = c.accept(s.asInstanceOf[Valid])
          if (s.isValid) {
            lastValid = s
          }
        }
        lastValid
      } else {
        s
      }
    }
  }


  case class Times(c: Consumer, N: Int) extends Consumer {
    def accept(state: Valid): State = {
      var s = c.accept(state)
      if (s.isValid) {
        var n = 1
        var lastValid = s
        while (s.isValid && s.hasInput && n < N) {
          s = c.accept(s.asInstanceOf[Valid])
          if (s.isValid) {
            lastValid = s
            n += 1
          }
        }
        if (n == N) lastValid else state.reject("Not enough repetitions")
      } else {
        s
      }
    }
  }

  def choice(first: Consumer, rest: Consumer*): Consumer = {
    rest.foldLeft(first) { case (acc, c) =>
      acc :| c
    }
  }

  import Token._

  def char(c: Char) = AcceptChar(c)
  def str(s: String) = AcceptString(s)
  def any(c: Char*) = AcceptAny(c)
  def anyBut(c: Char*) = AcceptAnyBut(c)
  val whitespace = AcceptAny(' ', '\r', '\n', '\t').*
  val skipWhitespace = whitespace // :| Nop
  val digit = AcceptAny('0' to '9')
  val exponent = (char('e') :| 'E') :> (char('+') :| '-' :| Nop) :> digit.*
  val decimals = (char('.') :> digit.*)
  val number_ = digit.* :> (decimals :| Nop) :> (exponent :| Nop)
  val alpha = AcceptAny('a' to 'z') :| AcceptAny('A' to 'Z')
  val alphaNumeric = alpha :| digit
  val ident_ = alpha :> ((alphaNumeric :| '_').* :| Nop)
  def keyword(k: String) = str(k) :> Not(ident_)

  val singleLineComment = str("//") :> anyBut('\n').*
  val multiLineComment = str("/*") :> (anyBut('*') :| (char('*') :> anyBut('/'))).* :| str("*/")

  val ident = ident_.!(Ident)

  val numberLit = number_.!((s: String, pos: Int) => NumberLit(s.toDouble, pos))

  val date_ = digit(4) :> '-' :> digit(2) :> '-' :> digit(2)
  val timezoneOffset_ = (char('+') :| '-') :> digit(2) :> ':' :> digit(2)
  val time_ = char('T') :> digit(2) :> ':' :> digit(2) :> ':' :> digit(2) :> (char('Z') :| timezoneOffset_ :| Nop)
  val datetime_ = date_ :> (time_ :| Nop)
  val dateLit = datetime_.!!((s, pos) => Date.fromString(s).map { d => DateLit(d, pos) })

  val stringEscapeSeq = char('\\') :> any('r', 'n', 't', '"', '\'', '\\')
  val stringLitDq = char('"') :> (anyBut('\\', '"') :| stringEscapeSeq).* :> char('"')
  val stringLitSq = char('\'') :> (anyBut('\\', '\'') :| stringEscapeSeq).* :> char('\'')
  val stringLit = (stringLitDq :| stringLitSq).!((s, pos) => StringLit(s.drop(1).dropRight(1), pos))

  val nullLit = keyword("null").!(NullLit)
  val trueLit = keyword("true").!(TrueLit)
  val falseLit = keyword("false").!(FalseLit)

  val bar = char('|').!(Bar)
  val lparen = char('(').!(LParen)
  val rparen = char(')').!(RParen)
  val comma = char(',').!(Comma)

  val opNot = char('!').!(OpNot)
  val opPlus = char('+').!(OpPlus)
  val opMinus = char('-').!(OpMinus)
  val opMultiply = char('*').!(OpMultiply)
  val opDivide = char('/').!(OpDivide)
  val opAssign = char('=') ! OpAssign
  val opEqual = str("==") ! OpEqual
  val opNotEqual = str("!=") ! OpNotEqual
  val opLess = char('<') ! OpLess
  val opLessEq = str("<=") ! OpLessEq
  val opGreater = char('>') ! OpGreater
  val opGreaterEq = str(">=") ! OpGreaterEq
  val opContains = keyword("contains") ! (OpContains)
  val opNotContains = keyword("!contains") ! (OpNotContains)
  val opAnd = keyword("and") ! (OpAnd)
  val opOr = keyword("or") ! (OpOr)

  val op = choice(
    opPlus, opMinus,
    opMultiply, opDivide,
    opEqual, opNotEqual,
    opAssign,
    opLessEq, opLess,
    opGreaterEq, opGreater,
    opContains, opNotContains,
    opNot,
    opAnd, opOr,
  )

  val tokens = choice(
    singleLineComment, multiLineComment, skipWhitespace, op,
    nullLit, trueLit, falseLit, ident, dateLit, numberLit, stringLit,
    bar, lparen, rparen, comma,
  ).*

  sealed trait Result
  final case class Success(tokens: Iterable[Token]) extends Result
  final case class Failure(error: String, loc: Location) extends Result

  def valid(input: String) = Valid(Input(input, pos = 0), Vector.empty)

  def lex(input: String): Result = {
    val state: Valid = valid(input)
    lexState(state) match {
      case Valid(input, tokens) if !input.hasInput =>
        Success(tokens)
      case Valid(input, tokens) =>
        Failure(s"Invalid token '${input.current()}'", Location(input.pos))
      case Reject(prev, message) => Failure(s"Lexing error: $message", Location(prev.input.pos))
    }
  }

  def lexState(state: Valid): State = {
    tokens.accept(state)
  }

}

